File: Google-Kick-Start/2017/Kick_Start_2017-D-1.py
# Sightseeing
# https://codingcompetitions.withgoogle.com/kickstart/round/0000000000201b77/0000000000201bfd
# DP, TLE but correct locally on test set 2

import math

INF = pow(10, 10)

# The max # of sightseeing if reach city i at sec start
def go(start, i):
    # print("start", start, "i", i)
    if i == N - 1:
        if start > T_F:
            return -INF
        else:
            return 0
    
    if (start, i) in dp:
        return dp[(start, i)]
    
    (S, F, D) = bus[i]
    # Go without sightseeing city i
    x = max(math.ceil((start - S) / F), 0)
    # print("option 1", x)
    option1 = go(S + x * F + D, i + 1)
    # Go with sightseeing city i
    # print("option 2", x)
    x = max(math.ceil((start + T_S - S) / F), 0)
    option2 = 1 + go(S + x * F + D, i + 1)

    dp[(start, i)] = max(option1, option2)
    return dp[(start, i)]


T = int(input())
for t in range(T):
    [N, T_S, T_F] = [int(x) for x in input().split()]
    bus = []
    for _ in range(N-1):
        [s, f, d] = [int(x) for x in input().split()]
        bus.append((s, f, d))

    dp = {}

    ans = go(0, 0)
    if ans < 0:
        print(f"Case #{t + 1}: IMPOSSIBLE")
    else:
        print(f"Case #{t + 1}: {ans}")

File: Google-Kick-Start/2017/Kick_Start_2017-D-1.cpp
// Sightseeing
// https://codingcompetitions.withgoogle.com/kickstart/round/0000000000201b77/0000000000201bfd
// DP, TLE but correct locally on test set 2

#include <bits/stdc++.h>

typedef long long ll;

using namespace std;

ll INF = pow(10, 10);

ll N, T_S, T_F;
vector< vector<ll> > bus;
map< pair<ll, int>, ll> dp;

// The max # of sightseeing if reach city i at sec start
ll go(ll start, int i) {
    if (i == N - 1) {
        if (start > T_F)
            return -INF;
        else
            return 0;
    }

    pair<ll, int> key = make_pair(start, i);
    if (dp.find(key) != dp.end()) {
        return dp[key];
    }
    
    // Go without sightseeing city i
    int x = max((int)ceil((double)(start - bus[i][0]) / bus[i][1]), 0);
    ll option1 = go(bus[i][0] + x * bus[i][1] + bus[i][2], i + 1);
    // Go with sightseeing city i
    x = max((int)ceil((double)(start + T_S - bus[i][0]) / bus[i][1]), 0);
    ll option2 = 1 + go(bus[i][0] + x * bus[i][1] + bus[i][2], i + 1);

    dp[key] = max(option1, option2);
    return dp[key];

}

int main() {
    int T;
    cin >> T;
    for (int t = 0; t < T; t++) {
        cin >> N >> T_S >> T_F;
        bus.clear();
        dp.clear();
        for (int i = 0; i < N - 1; i++) {
            vector<ll> temp(3);
            cin >> temp[0] >> temp[1] >> temp[2];
            bus.push_back(temp);
        }

        ll ans = go(0, 0);
        if (ans < 0)
            printf("Case #%d: IMPOSSIBLE\n", t + 1);
        else
            printf("Case #%d: %lld\n", t + 1, ans);
    }
}
